### ISSUE TITLE
I can't use pb file

### ISSUE THREAD

**Issue by A-Nariman** [thtrieu/darktf#881]
I trained my model from scratch without pre-trained weights. Using pb file cannot detect the objects even with threshold 0.1 while the checkpoints are fine and the model can detect the objects perfectly.

I read the all of issues about pb file but there was no solution. I would like to know what the problem is? Is it Darknet or Tensorflow? or something else?

I used the following commands:
1) Training:
------------------------------------------------------------------------------------------------
flow --model cfg/train.cfg --train --annotation train/xml --dataset train/png
------------------------------------------------------------------------------------------------
2) Saving pb file:
------------------------------------------------------------------------------------------------
flow --model cfg/train.cfg --load -1 --savepb
------------------------------------------------------------------------------------------------
3) Testing with checkpoints:
------------------------------------------------------------------------------------------------
flow --imgdir sample_img/ --model cfg/train.cfg --load -1
------------------------------------------------------------------------------------------------
4) Testing with pb file:
------------------------------------------------------------------------------------------------
flow --pbLoad built_graph/train.pb --metaLoad built_graph/train.meta --imgdir sample_img/

**Comment by zhangwenwen** [comment]
solved？

**Comment by amirpr24** [comment]
Yes, I trained my model with a pre-trained weight and worked. Without pre-trained weight does not work!!!

**Comment by braunyyy** [comment]
I've got the same problem, have you got an idea how to fix this?

**Comment by jkstyle2** [comment]
Can we not train without any pretrained weights? I've tried to predict some objects after training from the scratch, but I continue to fail detecting. Any suggestion?

**Comment by amankumarjain** [comment]
> Yes, I trained my model with a pre-trained weight and worked. Without pre-trained weight does not work!!!

how to do without pretrained weight?

**Comment by phquanta** [comment]
I was able to get functional pb file with freezing the graph from checkpoint with some customization. 
Problem is it is not tflite friendly now.


#!/usr/bin/python
# -*- coding: utf-8 -*-
import os
import argparse
import numpy as np

import tensorflow as tf

# The original freeze_graph function
# from tensorflow.python.tools.freeze_graph import freeze_graph

dir = os.path.dirname(os.path.realpath(__file__))


def freeze_graph(model_dir, output_node_names):
    """Extract the sub graph defined by the output nodes and convert 
    all its variables into constant 
    Args:
        model_dir: the root folder containing the checkpoint state file
        output_node_names: a string, containing all the output node's names, 
                            comma separated
    """

    if not tf.gfile.Exists(model_dir):
        raise AssertionError("Export directory doesn't exists. Please specify an export directory: %s"
                              % model_dir)

    if not output_node_names:
        print( 'You need to supply the name of a node to --output_node_names.')
        return -1

    # We retrieve our checkpoint fullpath

    checkpoint = tf.train.get_checkpoint_state(model_dir)
    input_checkpoint = checkpoint.model_checkpoint_path

    # We precise the file fullname of our freezed graph

    absolute_model_dir = '/'.join(input_checkpoint.split('/')[:-1])
    output_graph = absolute_model_dir + '/frozen_modelNew.pb'

    # We clear devices to allow TensorFlow to control on which device it will load operations

    clear_devices = True

    # We start a session using a temporary fresh Graph

    with tf.Session(graph=tf.Graph()) as sess:

        # We import the meta graph in the current default Graph

        saver = tf.train.import_meta_graph(input_checkpoint + '.meta',
                clear_devices=clear_devices)
        saver.restore(sess, input_checkpoint)
        
        #with tf.Graph().as_default() as temp_graph:
        print("JEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEee")
 #       for node in tf.get_default_graph().as_graph_def().node:
 #                   print(node.name)
        print("PPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPP")
#        with tf.Graph().as_default() as temp_graph:
        #graph_def=tf.get_default_graph().as_graph_def()
        #for node in graph_def.node:
        #               print(node.name)
            
        tf.import_graph_def(tf.get_default_graph().as_graph_def(),
                          input_map={"0-convolutional/is_training/input:0": tf.constant(False)})
        
            
        #print("Temp Graph:",temp_graph)
     
        # We use a built-in TF helper to export variables to constants

        output_graph_def = \
            tf.graph_util.convert_variables_to_constants(sess,
                tf.get_default_graph().as_graph_def(),
#                temp_graph.as_graph_def(),
                output_node_names.split(','))  # The session is used to retrieve the weights
                                               # The graph_def is used to retrieve the nodes
                                               # The output node names are used to select the usefull nodes

        #is_training = sess.graph.get_tensor_by_name("0-convolutional/is_training:0")
       
                                       

    # fix nodes

        for node in output_graph_def.node:
#            print(node)
            if node.op == 'RefSwitch':
                node.op = 'Switch'
                for index in range(len(node.input)):
                    if 'moving_' in node.input[index]:
                        node.input[index] = node.input[index] + '/read'
            elif node.op == 'AssignSub':
                node.op = 'Sub'
                if 'use_locking' in node.attr:
                    del node.attr['use_locking']
            
           #print("ISTRAINING", is_training_ph)
#        with tf.Graph().as_default() as temp_graph:
#            tf.import_graph_def(output_graph_def,
#                      input_map={is_training_ph.name: tf.constant(False)})           

        # Finally we serialize and dump the output graph to the filesystem
#        saver = tf.train.Saver()
        fd=np.ones((1,224,320,3),dtype=float)
        X = tf.placeholder(dtype=tf.float32, shape=[1,224,320,3])
        print(fd,X)
          #X = tf.placeholder(dtype=tf.float32, shape=[1,224,320,3])
        X= tf.get_default_graph().get_tensor_by_name("input:0")
        out= tf.get_default_graph().get_tensor_by_name("output:0")
        test = tf.get_default_graph().get_tensor_by_name("20-convolutional/biases:0")
        #arr=sess.run(out,feed_dict={X:fd})
        #print(arr)
        
        with tf.gfile.GFile(output_graph, 'wb') as f:
            f.write(output_graph_def.SerializeToString())
        tf.train.write_graph(output_graph_def, '.', 'tfdroid.pbtxt')      
        converter=tf.compat.v1.lite.TFLiteConverter.from_frozen_graph(output_graph,['input'],['output'])
        tflite_quant_model = converter.convert()
        open("PT.tflite","wb").write(tflite_quant_model)

        
        print ('%d ops in the final graph.' % len(output_graph_def.node))

    return output_graph_def


if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('--model_dir', type=str, default='',
                        help='Model folder to export')
    parser.add_argument('--output_node_names', type=str, default='',
                        help='The name of the output nodes, comma separated.'
                        )
    args = parser.parse_args()

    freeze_graph(args.model_dir, args.output_node_names)
